File: frontend/src/types/activity.ts
export type ActivitySummary = {
  id: string;
  provider: string;
  provider_activity_id: string;
  activity_type: string | null;
  started_at: string | null;
  duration_seconds: number | null;
  distance_meters: number | null;
  average_speed: number | null;
  average_heart_rate: number | null;
  max_heart_rate: number | null;
  average_cadence: number | null;
  elevation_gain: number | null;
  calories: number | null;
  created_at: string;
  updated_at: string;
};

export type ActivitySample = {
  timestamp: string;
  elapsed_seconds: number;
  distance_meters: number | null;
  heart_rate: number | null;
  speed: number | null;
  cadence: number | null;
  elevation: number | null;
};

export type ActivityDetail = ActivitySummary & {
  samples: ActivitySample[];
};

export type ActivityListQuery = {
  limit: number;
  offset: number;
  fromDate?: string;
  toDate?: string;
  activityType?: string;
};

export type ActivityListResponse = {
  items: ActivitySummary[];
  total: number;
  limit: number;
  offset: number;
  last_sync_at: string | null;
  activity_types: string[];
};

export type ActivitySyncResponse = {
  provider: string;
  created: number;
  updated: number;
  total: number;
  last_sync_at: string;
};

export type UnavailableMetric = {
  available: false;
  label: string;
  headline: string;
  note: string;
};

export type FiveKEstimateAvailable = {
  available: true;
  label: string;
  headline: string;
  note: string;
  estimated_seconds: number;
  qualifying_run_count: number;
};

export type EasyPaceAvailable = {
  available: true;
  label: string;
  headline: string;
  note: string;
  pace_seconds_per_km: number;
  comparable_pace_seconds_per_km: number;
  average_heart_rate: number;
  run_count: number;
  heart_rate_min: number;
  heart_rate_max: number;
};

export type AerobicEfficiencyAvailable = {
  available: true;
  label: string;
  headline: string;
  note: string;
  direction: "improving" | "stable" | "declining" | "not_enough_data";
  direction_label: string;
  score: number | null;
  relative_change_percent: number | null;
  qualifying_run_count: number;
};

export type FiveKEstimateMetric = FiveKEstimateAvailable | UnavailableMetric;
export type EasyPaceMetric = EasyPaceAvailable | UnavailableMetric;
export type AerobicEfficiencyMetric = AerobicEfficiencyAvailable | UnavailableMetric;

export type WeeklyVolume = {
  run_count: number;
  distance_meters: number;
  duration_seconds: number;
  period_start: string;
  period_end: string;
};

export type PaceHeartRatePoint = {
  activity_id: string;
  started_at: string;
  pace_seconds_per_km: number | null;
  average_heart_rate: number | null;
  distance_meters: number | null;
};

export type DashboardResponse = {
  weekly: WeeklyVolume;
  recent_activities: ActivitySummary[];
  pace_heart_rate_trend: PaceHeartRatePoint[];
  five_k_estimate: FiveKEstimateMetric;
  easy_pace: EasyPaceMetric;
  aerobic_efficiency: AerobicEfficiencyMetric;
};

export type EasyRunningPoint = {
  activity_id: string | null;
  started_at: string;
  pace_seconds_per_km: number;
  heart_rate: number;
  comparable_pace_seconds_per_km: number;
  distance_meters: number;
};

export type EasyRunningResponse = {
  heart_rate_min: number;
  heart_rate_max: number;
  run_count: number;
  distance_meters: number;
  average_pace_seconds_per_km: number | null;
  average_heart_rate: number | null;
  comparable_pace_seconds_per_km: number | null;
  headline: string;
  note: string;
  points: EasyRunningPoint[];
};

export type AerobicPoint = {
  activity_id: string | null;
  started_at: string;
  score: number;
  pace_seconds_per_km: number;
  heart_rate: number;
};

export type AerobicEfficiencyResponse = {
  metric: AerobicEfficiencyMetric;
  points: AerobicPoint[];
};

export type TrendActivityPoint = {
  activity_id: string | null;
  started_at: string;
  pace_seconds_per_km: number | null;
  average_heart_rate: number | null;
  distance_meters: number | null;
  comparable_pace_seconds_per_km: number | null;
};

export type WeeklyTrendPoint = {
  week_start: string;
  distance_meters: number;
  run_count: number;
};

export type TrendsResponse = {
  range_key: string;
  period_start: string | null;
  period_end: string;
  heart_rate_min: number;
  heart_rate_max: number;
  points: TrendActivityPoint[];
  weekly: WeeklyTrendPoint[];
};

export type TrendRangeKey = "4w" | "8w" | "3m" | "6m" | "1y" | "all";

File: backend/app/schemas/activity.py
"""Activity API schemas. ORM models are never returned directly."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ActivitySamplePublic(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    timestamp: datetime
    elapsed_seconds: int
    distance_meters: float | None
    heart_rate: int | None
    speed: float | None
    cadence: float | None
    elevation: float | None


class ActivitySummary(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: UUID
    provider: str
    provider_activity_id: str
    activity_type: str | None
    started_at: datetime | None
    duration_seconds: int | None
    distance_meters: float | None
    average_speed: float | None
    average_heart_rate: int | None
    max_heart_rate: int | None
    average_cadence: float | None
    elevation_gain: float | None
    calories: float | None
    created_at: datetime
    updated_at: datetime


class ActivityDetail(ActivitySummary):
    samples: list[ActivitySamplePublic]


class ActivitySampleCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    timestamp: datetime
    elapsed_seconds: int = Field(ge=0)
    distance_meters: float | None = Field(default=None, ge=0)
    heart_rate: int | None = Field(default=None, ge=0, le=400)
    speed: float | None = Field(default=None, ge=0)
    cadence: float | None = Field(default=None, ge=0)
    elevation: float | None = None


class ActivityCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    provider: str = Field(min_length=1, max_length=32)
    provider_activity_id: str = Field(min_length=1, max_length=128)
    activity_type: str | None = Field(default=None, max_length=64)
    started_at: datetime | None = None
    duration_seconds: int | None = Field(default=None, ge=0)
    distance_meters: float | None = Field(default=None, ge=0)
    average_speed: float | None = Field(default=None, ge=0)
    average_heart_rate: int | None = Field(default=None, ge=0, le=400)
    max_heart_rate: int | None = Field(default=None, ge=0, le=400)
    average_cadence: float | None = Field(default=None, ge=0)
    elevation_gain: float | None = Field(default=None, ge=0)
    calories: float | None = Field(default=None, ge=0)
    samples: list[ActivitySampleCreate] = Field(default_factory=list)


class ActivityListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[ActivitySummary]
    total: int
    limit: int
    offset: int
    last_sync_at: datetime | None


class ActivitySyncResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    provider: str
    created: int
    updated: int
    total: int
    last_sync_at: datetime
